### DISCUSSION TITLE
Reuse V2Ray instance in the manager

### DISCUSSION THREAD

**Discussion by v1rtl** [shadowsocks/shadowsocks-rust#1665] [Q&A]
I have a setup with SS manager where I dynamically add configs

all of them need V2Ray as an outbound proxy. I use it through v2ray-plugin.

however, there's a drawback — for each new configuration, a new v2ray instance opens

is it possible to only open one instance of the plugin and reuse it for all others?

**Comment by zonyitoo** [comment]
You can make your own plugin, which communicates to the standalone v2ray instance with IPC (sockets, pipes, shared memories, ...).

**Reply by v1rtl** [reply]
is it possible for this plugin to be compatible with V2Ray on the client? Do both of the implementations have to change? I'm not sure how exactly this would work

**Reply by zonyitoo** [reply]
Of course, it will be compatible with v2ray.

1. There must be a "plugin" process, that starts with each individual "local" and "server" instance, which implements SIP003 to communicate with sslocal and ssserver. It will read data from shadowsocks' service, and then bypasses all of them to the standalone v2ray service.
2. The v2ray service will see each individual connections from the "plugin" processes just like normal clients. It could get the destination addresses from each connections, then establish v2ray tunnel to the v2ray server.
3. v2ray server receives and then bypasses received connections to the server's "plugin" process. It will then bypasses to ssserver, which will then decrypt and handle the shadowsocks packets.

**Reply by v1rtl** [reply]
I see, thank you for the advice! I will try my skills in writing a modified V2Ray plugin that would bind to a single port on the server.

**Reply by v1rtl** [reply]
So, the client instance has to be modified as well? I'm not sure if I fully follow

not sure what "bypassing ssserver" would mean in this context as well.

**Reply by zonyitoo** [reply]
```
CLIENT -> sslocal -> PLUGIN
                       |
                       v
                  v2ray singleton
                       |
                       v
                     PLUGIN -> ssserver -> REMOTE TARGET
```

**Reply by v1rtl** [reply]
No I meant how can I bypass creating a new instance and connect to a previously created one?

**Reply by zonyitoo** [reply]
The `v2ray singleton` is a standalone process in your environment. Every local instance will create a new `PLUGIN` instance, which connects to the `v2ray singleton`.
